AWS EC2 Instance tag filter

prune998 opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: https://github.com/cilium/cilium/issues/18239

christarazi reviewed (changes requested):
Why is the subnet tag filter not enough to do this job? AFAIU, interfaces are attached to subnets, so filtering out certain subnets allows you to filter out interfaces as well.

gandro commented on operator/option/config.go:
```diff
@@ -453,6 +459,12 @@ func (c *OperatorConfig) Populate() {
 
+	if m, err := command.GetStringMapStringE(viper.GetViper(), IPAMInstancesTags); err != nil {
+		log.Fatalf("unable to parse %s: %s", IPAMInstancesTags, err)
+	} else if len(m) != 0 {
```
We had to remove this check, see #19172 for how this now looks on current master
```suggestion
	} else {
```

prune998 replied:
I just rebased the PR on master

gandro commented on pkg/aws/ec2/ec2.go:
```diff
@@ -96,6 +98,21 @@ func NewSubnetsFilters(tags map[string]string, ids []string) []ec2_types.Filter
 	return filters
 }
 
+// NewInstancsFilters transforms a map of tags and values
```
```suggestion
// NewInstancesFilters transforms a map of tags and values
```

prune998 replied:
changed from rebase on master

gandro commented on pkg/aws/ec2/ec2.go:
```diff
@@ -142,6 +159,61 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context, subnets ipamType
+
+	paginator := ec2.NewDescribeInstancesPaginator(c.ec2Client, instanceAttrs)
+	for paginator.HasMorePages() {
+		c.limiter.Limit(ctx, "DescribeNetworkInterfacesFromInstances")
```
The name here refers to the AWS API:
```suggestion
		c.limiter.Limit(ctx, "DescribeInstances")
```

prune998 replied:
updated

gandro commented on pkg/aws/ec2/ec2.go:
```diff
@@ -142,6 +159,61 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context, subnets ipamType
+		c.limiter.Limit(ctx, "DescribeNetworkInterfacesFromInstances")
+		sinceStart := spanstat.Start()
+		output, err := paginator.NextPage(ctx)
+		c.metricsAPI.ObserveAPICall("DescribeNetworkInterfacesFromInstances", deriveStatus(err), sinceStart.Seconds())
```
```suggestion
		c.metricsAPI.ObserveAPICall("DescribeInstances", deriveStatus(err), sinceStart.Seconds())
```

prune998 replied:
updated

gandro commented on pkg/aws/ec2/ec2.go:
```diff
@@ -217,7 +289,14 @@ func parseENI(iface *ec2_types.NetworkInterface, vpcs ipamTypes.VirtualNetworkMa
+	var networkInterfaces []ec2_types.NetworkInterface
+	var err error
+
+	if len(c.instancesFilters) > 0 {
```
So this preserves the existing subnet tag filter behavior if no instance filter is set. That is great, because it won't surprise anyone who upgrades with a new behavior.

But I do wonder if we maybe should just always call into `describeNetworkInterfacesFromInstances` here to avoid the additional code complexity. The problem with the subnet tag filter is that it only works here if the subnet tag filter matches the subnet of default ENI on the fetched instances, which is a bit of a surprising behavior. Therefore I do wonder if we should just get rid of the surprising behavior here, and ask users to also set a instance filter if they are concerned about scalability of the query.

It would however require large-scale user to define a instance tag filter before upgrading to avoid overly large result sets, so not sure if that's considered a breaking change. What are the opinions on this? (e.g. @christarazi?)

prune998 replied:
Yeah, I'm curious about your (Cilium maintainers) view. 
I built up the PR this way to avoid any breaking change. 
The truth is, `instance tag filter` will work better than `subnet tag filter` in any complex scenario. My take is that we should deprecate the `subnet tag filter` in a coming version...
That said, we could also bring both filters together in the same function and do the check there...

christarazi replied:
@gandro It sounds like a breaking change to me, since users are definitely configuring Cilium specifically around this. 

@prune998 Let's add an entry to the upgrade docs and have docs around this in general to help communicate what's being done in this PR and what's recommended moving forward.

gandro replied:
I thought about this some more, and I strongly believe now we should change the behavior of the `subnet tag filter` and not apply it in `GetInstances`. We've now had multiple users reporting it breaking their cluster due to it's behavior being very surprising. Therefore I think changing the behavior (while technically a breaking change) will actually make the life of most users easier.

Therefore, I'm proposing to always call `describeNetworkInterfacesFromInstances` here and delete the old `describeNetworkInterfaces` function. Users who have scalability concerns can use the new `instance tag filter`.

prune998 replied:
@gandro, while I agree with you, do we have to do this now in this PR or do you want to wait and deprecate the `subnet tag filter` later on ?
We could add the new `instance tag filter` now and document deprecation of the`subnet tag filter`, and remove it on the following release ? 
Let me know and I'll update the code.

prune998 replied:
I rebased, corrected, and added the new option into the Helm chart + the upgrade docs as a new parameter ; I also added a deprecation warning for the `subnet tag filter`.
Let me know what you think.

gandro replied:
Interesting. So I checked the source and `GetSubnets` also uses the `subnet tag filter` and I actually wonder now how this interacts with this PR:

So `GetSubnets` also uses the `subnet tag filter`
https://github.com/cilium/cilium/blob/7c1bb3592c46e6d3bc04d3eb8ff37aaf70f06dc3/pkg/aws/ec2/ec2.go#L299

The result gets passed to `GetInstances` here:
https://github.com/cilium/cilium/blob/440c896d10b29e110bb4cc3e708b1fc90c5b616a/pkg/aws/eni/instances.go#L168) 

Which is then used to populate the ENI information here:
https://github.com/cilium/cilium/blob/7c1bb3592c46e6d3bc04d3eb8ff37aaf70f06dc3/pkg/aws/ec2/ec2.go#L237
https://github.com/cilium/cilium/blob/7c1bb3592c46e6d3bc04d3eb8ff37aaf70f06dc3/pkg/aws/ec2/ec2.go#L194

I do wonder how that now interacts with the `instance tags filter` - if one specifies the `subnet tag filter` *and* an `instance tag filter`, then I think `describeNetworkInterfacesFromInstances` will return ENIs attatched to subnets _not_ matching of the `subnet tag filter`. Does this mean that the subnet information will not be populated?

I believe that if a user specifies a `subnet tag filter`, then I think we have to also use it to filter ENIs in `describeNetworkInterfacesFromInstances`. Otherwise, Cilium will consider them for IPAM, which would be wrong.

gandro replied:
Unless I'm misinterpreting the code, I think the `subnet tag filter` still serves a purpose (i.e. trimming the result set of `GetSubnets`). We should therefore make sure that the `instance tag filter` works well together with the `subnet tag filter`, instead of trying to deprecate and replace it.

prune998 replied:
`subnet` and `instance` tag filters are just limiting what information Cilium is managing. 
You actually want Cilium to know about ALL the ENIs of the instances that you're using in your cluster. This is what you get by using `instance tag filter`.
This is not guaranty if you use `subnet tag filter`. 
But I don't see any case where you would want to use both.

Cilium is using the value from `cilium-cni-configuration` ConfigMap to pick the subnet to use to add new IPs, not the data filtered by the `subnet tag filter` (as far as I know)
```
          "eni": {
            "subnet-tags":{
              "Usage":"pods"
            }
```

You're right that `subnet tag filter` is useful to limit the number of subnets Cilium will sync, but it is not relevant to use is to `GetInstances`

gandro replied:
>  You actually want Cilium to know about ALL the ENIs of the instances that you're using in your cluster. This is what you get by using instance tag filter.

I agree that we want all ENIs. However, for each ENI we will also require its subnet. Will your current implementation contain the subnet information extracted here?
https://github.com/cilium/cilium/blob/7c1bb3592c46e6d3bc04d3eb8ff37aaf70f06dc3/pkg/aws/ec2/ec2.go#L193-L198

If not, we still will have to fetch the related subnets manually and be smart about it.

prune998 replied:
`subnets` is coming from the `GetSubnets()` func, so it will be either all subnets of the VPC if there's no subnet filter, and some subnets if you use a filter.
My take here, is that the `subnet` filter was implemented as a quick way to limit the number of ENIs (and instances) to maintain. Usually, you have a lot of instances and a lot of ENIs/IPs, but not that many subnets in the same account. And even if it was a hundred subnets, i'm not sure it would have a huge impact.

Ideally, if a user wants to perfectly tune the AWS API search so Cilium would only maintain something that can be used, then they would define both a `subnet` filter so `describeSubnets()` would only grab useful subnets AND an `instance` filter to only grab instances of the cluster.
Of course, we could replace all that and get the list of subnets from the selected instances... but that would be a lot of changes, and I don't know how other providers (alibaba, google, azure) would behave....
 
That said, the outcome is that we should keep the `subnet` filter for `GetSubnets()`  and keep the `instance` filter for `GetInstances()`. And it will be up to the user to ensure the filters matches...

gandro replied:
> That said, the outcome is that we should keep the subnet filter for GetSubnets() and keep the instance filter for GetInstances(). And it will be up to the user to ensure the filters matches...

Yes, I think that's a reasonable outcome. So to summarize this means we won't deprecate or change the behavior of the `subnet tag filter` option, but make sure we document it properly.

I already started working on a docs PR in https://github.com/cilium/cilium/pull/19276, but it's intentionally kept separate from this PR so we can backport it to older versions.

gandro reviewed (changes requested):
Overall this looks very solid to me, thanks a lot for taking care of this!  :rocket: 

I have a few minor nits that need to be addressed, and an overall question about what the interaction with this and the old subnet tag filter should be.

christarazi commented on pkg/aws/ec2/ec2.go:
```diff
@@ -96,6 +98,21 @@ func NewSubnetsFilters(tags map[string]string, ids []string) []ec2_types.Filter
+// NewInstancsFilters transforms a map of tags and values
+// into a slice of ec2.Filter adequate to filter AWS Instances.
+func NewInstancesFilters(tags map[string]string) []ec2_types.Filter {
+	filters := make([]ec2_types.Filter, 0, len(tags)+1)
```
Nit: why is it necessary to preallocate with +1 length?

prune998 replied:
I guess I got that from the `NewSubnetsFilters` and stopped thinking. Removing it.

christarazi commented on pkg/aws/ec2/ec2.go:
```diff
@@ -142,6 +159,61 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context, subnets ipamType
 
+// describeNetworkInterfacesFromInstances lists all ENIs matching filtered EC2 instances
+func (c *Client) describeNetworkInterfacesFromInstances(ctx context.Context) ([]ec2_types.NetworkInterface, error) {
+	ENIsFromInstances := make(map[string]*struct{})
```
Nit: this doesn't need to be a pointer to a struct and can just be struct{}.
```suggestion
	ENIsFromInstances := make(map[string]struct{})
```

prune998 replied:
ok changed.

christarazi commented on pkg/aws/ec2/ec2.go:
```diff
@@ -142,6 +159,61 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context, subnets ipamType
+	var result []ec2_types.NetworkInterface
+
+	ENIPaginator := ec2.NewDescribeNetworkInterfacesPaginator(c.ec2Client, ENIAttrs)
+	for ENIPaginator.HasMorePages() {
```
Nit: The variables starting with `ENI` don't need to be capitalized since capitalization means exporting which means nothing inside a function scope.

prune998 replied:
done

christarazi commented on operator/flags.go:
```diff
@@ -35,6 +35,10 @@ func init() {
 
+	flags.StringToStringVar(&operatorOption.Config.IPAMInstancesTags, operatorOption.IPAMInstancesTags, operatorOption.Config.IPAMInstancesTags,
+		"EC2 Instances tags in the form of k1=v1,k2=v2 (multiple k/v pairs can also be passed by repeating the CLI flag")
+	option.BindEnv(operatorOption.IPAMInstancesTags)
```
I'm noticing that we are inconsistent between `Instances tags` and `Instance tags`. I propose we go with the latter everywhere including the flag name as well.

prune998 replied:
yeah, I tried to reflect the `subnet(s) tag(s)` naming, but I'll change it to all use the same as `Instance tags`

gandro commented on Documentation/operations/upgrade.rst:
```diff
@@ -334,6 +338,8 @@ Removed Options
 * The ``host-reachable-services-protos`` option (``.hostServices.protocols`` in
   Helm) was deprecated, and it will be removed in version 1.13.
+* The ``subnet-tags-filter`` option (``.eni.subnetTagsFilter`` in Helm) was deprecated
+  and replaced by the ``instance-tags-filter`` option. It will be removed in version 1.13.
```
So I think given our conclusion in https://github.com/cilium/cilium/pull/19181/files#r838850096 I think we want to un-deprecate this, right?

prune998 replied:
To be clear... we keep both option, BUT do we keep it as-is, where `instance filter` has precedence over `subnet filter` in `GetInstances()` ? 
(I guess we should keep it like that)

prune998 replied:
I updated the code to remove the deprecation notice

gandro replied:
> To be clear... we keep both option, BUT do we keep it as-is, where instance filter has precedence over subnet filter in GetInstances() ?

Yes, perfect, thanks!

gandro reviewed (commented):
Thanks a lot for the amazing work and discussions so far. I think this is starting to look very solid.

gandro reviewed (approved):
Awesome! Thanks a ton for taking care of this and helping with understanding all implications of this.

One last request: Could you squash the commits into one please? Thank you!

christarazi reviewed (commented):
Could you cleanup the commit msg text to remove all the previous commit msgs?

kaworu reviewed (approved):
Thanks for the PR @prune998, Helm related changes LGTM.

kaworu reviewed (changes requested):
Travis CI error looks legit:
> HINT: to fix this, run 'make -C Documentation update-cmdref'

prune998 committed:
adding AWS instance Tag filter

This PR is adding a new command line option `instance-tags-filter`  to the Cilium-Operator to filter maintained AWS EC2 instances (and available ENIs) based on EC2 instance tags

Technically, if an instance tag filter is provided, instead of searching ALL the ENIs that exist in the AWS account, we:

- skip the `instance-subnet-filter` 
- search all instances matching the tags
- grab all the ENIS of the selected instances

The code will search for matching instances and will grab all instance ENIS (giving us `ec2_types.InstanceNetworkInterface`) 
 and will then search ENIs to get `ec2_types.NetworkInterface` that is needed down the line.

Signed-off-by: Sebastien Prune THOMAS <prune@lecentre.net>
